### PR TITLE
implimented requestpasswordreset endpoint to receive either email o…

### PR DESCRIPTION
**Pull request by fanyui**
implimented requestpasswordreset endpoint to receive either email or username and trigger the creation of an activation key for the user which will be sent via email to the user provided he / she owns a valid email.

### CODE REVIEW

**Comment by fanyui** [comment]
@wluyima  here is a pull request for the request for password reset endpoint.

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.http.HttpStatus;
+import org.springframework.stereotype.Controller;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@Controller
+@RequestMapping(value = "/rest/" + RestConstants.VERSION_1)
```
Comment (wluyima) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController.java]: You are missing the resource name passwordreset

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java [hunk 2]
```diff
@@ -12,59 +12,47 @@
...
 		
-		if (body.containsKey("email")) {
-			String email = body.get("email");
-			user = userService.getUserByEmail(email);
-		}
-		else if (body.containsKey("username")) {
-			String userName = body.get("username");
-			user = userService.getUserByUsername(userName);
+		if ((user = userService.getUserByEmail(usernameOrEmail)) == null) {
+			user = userService.getUserByUsername(usernameOrEmail);
```
Comment (wluyima) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: I think in openmrs-core, you will need to rename this method to `getUserByUsernameOrEmail` just for clarity of what it does

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java [hunk 3]
```diff
@@ -12,59 +12,47 @@
...
-		
-		String token = body.get("token");
-		
-		try {
-			userService.changeUserPasswordUsingActivationKey(token, newPassword);
-		}
-		catch (InvalidActivationKeyException ex) {
-			ex.printStackTrace();
-		}
+		String activationKey = body.get("token");
```
Comment (wluyima) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: should be activationKey

**Comment by wluyima** [comment]
Where are the unit tests? Please get into the habit of always including unit tests for all your changes

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+	
+	@RequestMapping(method = RequestMethod.POST)
+	@ResponseStatus(HttpStatus.OK)
+	public void requestPasswordReset(@RequestBody Map<String, String> body) {
+		User user = null;
+		String usernameOrEmail = body.get("usernameOrEmail");
+		if ((user = userService.getUserByEmail(usernameOrEmail)) == null) {
+			user = userService.getUserByUsername(usernameOrEmail);
+		}
+		//user = userService.getUserByEmail(usernameOrEmail);
```
Comment (wluyima) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: Please don't leave behind commented out code, just remove it

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+		if (user != null) {
+			userService.setUserActivationKey(user);
+		}
+	}
+	
+	@RequestMapping(value = "/{activationkey}", method = RequestMethod.POST)
+	@ResponseBody
+	public ResponseEntity<String> resetPassword(@PathVariable("activationkey") String activationkey,
```
Comment (wluyima) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: We don't return ResponseEntity, we ONLY return SimpleObject

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+@Controller
+@RequestMapping(value = "/rest/" + RestConstants.VERSION_1 + "/passwordreset")
+public class PasswordResetController2_2 extends BaseRestController {
+	
+	@Qualifier("userService")
+	@Autowired
+	private UserService userService;
+	
+	@RequestMapping(method = RequestMethod.POST)
+	@ResponseStatus(HttpStatus.OK)
```
Comment (wluyima) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: This should be at the class level

Comment (wluyima) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: Ignore this, I think keep them at the method level so that you don't affect methods from the superclass

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+		}
+	}
+	
+	@RequestMapping(value = "/{activationkey}", method = RequestMethod.POST)
+	@ResponseBody
+	public ResponseEntity<String> resetPassword(@PathVariable("activationkey") String activationkey,
+	        @RequestBody Map<String, String> body) {
+		String newPassword = body.get("newPassword");
+		if (userService.getUserByActivationKey(activationkey) == null) {
+			return new ResponseEntity<String>("Invalid Activation Key", HttpStatus.BAD_REQUEST);
```
Comment (wluyima) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: You should instead throw an exception

Comment (wluyima) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: Possibly a NPE

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+			userService.setUserActivationKey(user);
+		}
+	}
+	
+	@RequestMapping(value = "/{activationkey}", method = RequestMethod.POST)
+	@ResponseBody
+	public ResponseEntity<String> resetPassword(@PathVariable("activationkey") String activationkey,
+	        @RequestBody Map<String, String> body) {
+		String newPassword = body.get("newPassword");
+		if (userService.getUserByActivationKey(activationkey) == null) {
```
Comment (wluyima) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: This check is already done userService.changePasswordUsingActivationKey(activationkey) so don't do it here again, it's the reason why you added the new exception class, remember?

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+		}
+		else {
+			userService.changePasswordUsingActivationKey(activationkey, newPassword);
+			return new ResponseEntity<String>(HttpStatus.OK);
+		}
+	}
+	
+	@RequestMapping(value = "/{activationkey}", method = RequestMethod.GET)
+	@ResponseBody
+	public ResponseEntity<String> verifyActivationKey(@PathVariable("activationkey") String activationkey) {
```
Comment (wluyima) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: Why did you add this method?

File: omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+	private UserService userService;
+	
+	@Before
+	public void setup() {
+		userService = Context.getUserService();
+		messages = Context.getMessageSourceService();
+	}
+	
+	@Test
+	public void requestPasswordReset_shouldFailIfUsernameOrEmailIsBlanck() throws Exception {
```
Comment (wluyima) [inline omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java]: Why do you need this test in the weblayer? The logic is in the service layer, you don't need to duplicate the test here

File: omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+		String usernameOrEmail = "  ";
+		
+		expectedException.expect(APIException.class);
+		expectedException.expectMessage(messages.getMessage("error.email.notNullOrBlank"));
+		handle(newPostRequest(RESET_PASSWORD_URI, "{\"usernameOrEmail\":\"" + usernameOrEmail + "\"}"));
+		
+	}
+	
+	@Test
+	public void requestPasswordReset_shouldFailIfUsernameOrEmailIsEmptyl() throws Exception {
```
Comment (wluyima) [inline omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java]: This test isn't necessary in the web layer too

File: omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+		u.getPerson().setGender("M");
+		userService.createUser(u, "Openmr5xy");
+		String usernameOrEmail = "";
+		
+		expectedException.expect(APIException.class);
+		expectedException.expectMessage(messages.getMessage("error.email.notNullOrBlank"));
+		handle(newPostRequest(RESET_PASSWORD_URI, "{\"usernameOrEmail\":\"" + usernameOrEmail + "\"}"));
+		
+	}
+
```
Comment (wluyima) [inline omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java]: Where is the test for a valid password reset request?

File: omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+	@Test
+	public void resetPassword_shouldFailWith400BadRequestIfActivationKeyIsInvalid() throws Exception {
+		String activationkey = "wrongActivationKey12";
+		String newPassword = "newPassword9";
+		MockHttpServletResponse response = handle(newPostRequest(RESET_PASSWORD_URI + "/" + activationkey,
+		    "{\"newPassword\":\"" + newPassword + "\"}"));
+		assertEquals(400, response.getStatus());
+		
+	}
+}
```
Comment (wluyima) [inline omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java]: Where is the valid test for a valid change password request?

File: omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+	
+	@Test
+	public void verifyActivationKey_shouldFailWith400BadRequestIfActivationKeyIsInvalid() throws Exception {
+		String activationKey = "wrongActivationKey12";
+		MockHttpServletResponse response = handle(newGetRequest(RESET_PASSWORD_URI + "/" + activationKey));
+		assertEquals(400, response.getStatus());
+	}
+	
+	@Test
+	public void resetPassword_shouldFailWith400BadRequestIfActivationKeyIsInvalid() throws Exception {
```
Comment (wluyima) [inline omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java]: This behavior is in the service layer, you don't need to duplicate the test here?

**Comment by wluyima** [comment]
I believe you only need  2 tests, one for the request password reset and the other for the change password request, the ones you added don't make sense since they are testing logic that's actually in the service layer that you already added to core anyways.

File: omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java (hunk 12 above)
Comment (fanyui) [reply omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java]: @wluyima activation key expires so am having trouble writing a test for successfully verifying and reseting the password. please for some insight on how to do it

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java (hunk 9 above)
Comment (fanyui) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: @wluyima The sequence diagram Burke presented shows wee need a method to verify activationKey So this method is for that

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+	@Qualifier("userService")
+	@Autowired
+	private UserService userService;
+	
+	@RequestMapping(method = RequestMethod.POST)
+	@ResponseStatus(HttpStatus.OK)
+	public void requestPasswordReset(@RequestBody Map<String, String> body) {
+		User user = null;
+		String usernameOrEmail = body.get("usernameOrEmail");
+		if ((user = userService.getUserByEmail(usernameOrEmail)) == null) {
```
Comment (wluyima) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: This code looks vague, you assign user in the if clause, I think you don't have to do the assignment again inside the if clause

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+			response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+			return RestUtil.wrapErrorResponse(exception, "Password reset failed");
+		}
+		
+		return resp;
+	}
+	
+	@RequestMapping(value = "/{activationkey}", method = RequestMethod.GET)
+	@ResponseBody
+	public SimpleObject verifyActivationKey(@PathVariable("activationkey") String activationkey, HttpServletRequest request,
```
Comment (wluyima) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: Why is this method needed? You should not expose a rest endpoint to verify the activationKey

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+			user = userService.getUserByUsername(usernameOrEmail);
+		}
+		if (user != null) {
+			userService.setUserActivationKey(user);
+		}
+	}
+	
+	@RequestMapping(value = "/{activationkey}", method = RequestMethod.POST)
+	@ResponseBody
+	public SimpleObject resetPassword(@PathVariable("activationkey") String activationkey,
```
Comment (wluyima) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: Why do you need a return value anyways? I think this method should not return a value.

File: omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+	private UserService userService;
+	
+	@Before
+	public void setup() {
+		userService = Context.getUserService();
+		messages = Context.getMessageSourceService();
+	}
+	
+	@Test
+	public void verifyActivationKey_shouldFailWith400BadRequestIfActivationKeyIsInvalid() throws Exception {
```
Comment (wluyima) [inline omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java]: Where are the tests that test the setting a user activation key and changing the password?

This test won't be needed, because you don't need an endpoint to verify the activation key

File: omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+import org.openmrs.module.webservices.rest.web.v1_0.controller.RestControllerTestUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.mock.web.MockHttpServletResponse;
+
+public class PasswordResetController2_2Test extends RestControllerTestUtils {
+	
+	private static final String RESET_PASSWORD_URI = "passwordreset";
+	
+	private MessageSourceService messages;
```
Comment (wluyima) [inline omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java]: Why is this needed?

**Comment by fanyui** [comment]
Hi @wluyima  am having trouble writing the test for testing reseting of password as activation key is generated at runtime and usually expires and i wouldn't have access to a correct one to test with.

**Comment by wluyima** [comment]
Write them them the same way you wrote the service layer tests in core

**Comment by wluyima** [comment]
I'm a little confused, the activation key takes 10min to expire while a test takes milliseconds to run, there is no way the token can expire before the test completes.

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+		String usernameOrEmail = body.get("usernameOrEmail");
+		User user = userService.getUserByUsernameOrEmail(usernameOrEmail);
+		if (user != null) {
+			userService.setUserActivationKey(user);
+		}
+	}
+	
+	@RequestMapping(value = "/{activationkey}", method = RequestMethod.POST)
+	public void resetPassword(@PathVariable("activationkey") String activationkey,
+	        @RequestBody Map<String, String> body, HttpServletRequest request,
```
Comment (wluyima) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: If you never use the request parameter, remove it

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+	        HttpServletResponse response) {
+		String newPassword = body.get("newPassword");
+		
+		try {
+			userService.changePasswordUsingActivationKey(activationkey, newPassword);
+			response.setStatus(HttpServletResponse.SC_OK);
+		}
+		catch (InvalidActivationKeyException exception) {
+			response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
```
Comment (wluyima) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: You should instead rethrow a ValidationException and let the framework take care of it

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+	
+	@RequestMapping(value = "/{activationkey}", method = RequestMethod.POST)
+	public void resetPassword(@PathVariable("activationkey") String activationkey,
+	        @RequestBody Map<String, String> body, HttpServletRequest request,
+	        HttpServletResponse response) {
+		String newPassword = body.get("newPassword");
+		
+		try {
+			userService.changePasswordUsingActivationKey(activationkey, newPassword);
+			response.setStatus(HttpServletResponse.SC_OK);
```
Comment (wluyima) [inline omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: Status should be defined on the method

File: omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java [hunk 23]
```diff
@@ -0,0 +1,101 @@
...
+	private UserDAO dao;
+	
+	@Before
+	public void setup() {
+		userService = Context.getUserService();
+	}
+	
+	@Test
+	public void requestPasswordReset_shouldCreateUserActivationKeyGivenUsername() throws Exception {
+		User u = new User();
```
Comment (wluyima) [inline omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java]: Why don't you use an existing user?

File: omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java [hunk 24]
```diff
@@ -0,0 +1,101 @@
...
+		User createdUser = userService.createUser(u, "Openmr5xy");
+		assertNull(dao.getLoginCredential(createdUser).getActivationKey());
+		assertEquals(createdUser, userService.setUserActivationKey(createdUser));
+		handle(newPostRequest(RESET_PASSWORD_URI, "{\"usernameOrEmail\":\"" + "bwolfe" + "\"}"));
+		assertNotNull(dao.getLoginCredential(createdUser).getActivationKey());
+	}
+	
+	@Test
+	public void requestPasswordReset_shouldCreateUserActivationKeyGivenEmail() throws Exception {
+		User u = new User();
```
Comment (wluyima) [inline omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java]: Why don't you use an existing user?

File: omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java [hunk 25]
```diff
@@ -0,0 +1,101 @@
...
+		User createdUser = userService.createUser(u, "Openmr5xy");
+		assertNull(dao.getLoginCredential(createdUser).getActivationKey());
+		assertEquals(createdUser, userService.setUserActivationKey(createdUser));
+		handle(newPostRequest(RESET_PASSWORD_URI, "{\"usernameOrEmail\":\"" + "bwolfe@gmail.com" + "\"}"));
+		assertNotNull(dao.getLoginCredential(createdUser).getActivationKey());
+	}
+	
+	@Test
+	public void resetPassword_shouldResetUserPasswordIfActivationKeyIsCorrect() throws Exception {
+		User u = new User();
```
Comment (wluyima) [inline omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java]: Why don't you use an existing user?

File: omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java [hunk 26]
```diff
@@ -0,0 +1,101 @@
...
+		Long tokenTime = System.currentTimeMillis() + validTime;
+		LoginCredential credentials = dao.getLoginCredential(createdUser);
+		credentials
+		        .setActivationKey("b071c88d6d877922e35af2e6a90dd57d37ac61143a03bb986c5f353566f3972a86ce9b2604c31a22dfa467922dcfd54fa7d18b0a7c7648d94ca3d97a88ea2fd0:"
+		                + tokenTime);
+		dao.updateLoginCredential(credentials);
+		String newPassword = "newPasswordString123";
+		MockHttpServletResponse response = handle(newPostRequest(RESET_PASSWORD_URI + "/" + key, "{\"newPassword\":\""
+		        + newPassword + "\"}"));
+		assertEquals(200, response.getStatus());
```
Comment (wluyima) [inline omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java]: You should have an assertion that ensures that you can login with the new password

File: omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java (hunk 22 above)
Comment (wluyima) [reply omod-2.2/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2.java]: I think you might have missed one of my other comments, I did say re throw a ValidationException when InvalidActivationKeyException is caught, that should take care of it

File: omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java (hunk 23 above)
Comment (fanyui) [reply omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java]: @wluyima I have been checking and the last dataset that contained an existing user was defined since 1.8 so i am thinking declaring a new user here is appropriate.

File: omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java [hunk 27]
```diff
@@ -0,0 +1,84 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.mock.web.MockHttpServletResponse;
+
+public class PasswordResetController2_2Test extends RestControllerTestUtils {
+	
+	private static final String RESET_PASSWORD_URI = "passwordreset";
+	
+	@Rule
+	public ExpectedException expectedException = ExpectedException.none();
```
Comment (wluyima) [inline omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java]: Where are you using this?

File: omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java [hunk 28]
```diff
@@ -0,0 +1,79 @@
...
+	@Autowired
+	@Qualifier("userService")
+	private UserService userService;
+	
+	@Autowired
+	private UserDAO dao;
+	
+	@Before
+	public void setup() {
+		userService = Context.getUserService();
```
Comment (wluyima) [inline omod-2.2/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_2/PasswordResetController2_2Test.java]: This is redundant and a duplication because you are already autowiring it

**Comment by fanyui** [comment]
@wluyima  Please is there still anything left with this pull request?
